nilearn/image/resampling.py

```diff
@@ -452,27 +452,18 @@ def resample_img(img, target_affine=None, target_shape=None,
     if isinstance(target_shape, np.ndarray):
         target_shape = target_shape.tolist()
     target_shape = tuple(target_shape)
-    # For images with dimensions larger than 3D:
-    if len(data_shape) > 3:
-        # Iter in a set of 3D volumes, as the interpolation problem is
-        # separable in the extra dimensions. This reduces the
-        # computational cost
-        other_shape = data_shape[3:]
-        resampled_data = np.ndarray(list(target_shape) + other_shape,
-                                    order=order)
-
-        all_img = (slice(None), ) * 3
-
-        for ind in np.ndindex(*other_shape):
-            _resample_one_img(data[all_img + ind], A, A_inv, b, target_shape,
-                      interpolation_order,
-                      out=resampled_data[all_img + ind],
-                      copy=not input_img_is_string)
-    else:
-        resampled_data = np.empty(target_shape, data.dtype)
-        _resample_one_img(data, A, A_inv, b, target_shape,
+
+    # Code is generic enough to work for both 3D and 4D images
+    other_shape = data_shape[3:]
+    resampled_data = np.empty(list(target_shape) + other_shape,
+                              order=order, dtype=data.dtype)
+
+    all_img = (slice(None), ) * 3
```

Comment:
Well, blaming is the word here. It is far less readable than `[:, :, :]` and highly inefficient. There is an iterator to iterate on non-first dimensions, it is `np.nested_iters`. However this is not the fastest solution, transposing the axes is even faster. It has to be noted that, for that particular purpose, a `nditer` may be the faster solution because it would iter on both arrays at the same time.

Here is a small bench to convince you.

```python
import numpy as np
import time


a = np.random.random((20, 30, 40, 5, 6))

t0 = time.time()
for i in range(100):
    all_img = (slice(None), ) * 3
    other_shape = a.shape[3:]
    for ind in np.ndindex(*other_shape):
        a[all_img + ind].shape
print time.time() - t0

t0 = time.time()
for i in range(100):
    it0, it1 = np.nested_iters(a, [range(3, a.ndim), [0, 1, 2]], flags=['multi_index'])
    for x in it0:
        it1.itviews[0].shape
print time.time() - t0

t0 = time.time()
for i in range(100):
    other_shape = a.shape[3:]
    b = np.transpose(a, range(3, a.ndim) + [0, 1, 2])
    b = np.reshape(b, (-1,) + a.shape[:3])
    for x in b:
        x
print time.time() - t0
```

Output:
```
0.00588703155518
0.00144195556641
0.000957012176514
```